pulsar-zookeeper-utils/src/main/java/org/apache/pulsar/zookeeper/ZkIsolatedBookieEnsemblePlacementPolicy.java

```diff
@@ -156,34 +177,79 @@ private ZooKeeperCache getAndSetZkCache(Configuration conf) {
                 bookieToReplace, excludeBookies);
     }
 
-    private Set<BookieId> getBlacklistedBookies(int ensembleSize) {
-        Set<BookieId> blacklistedBookies = new HashSet<BookieId>();
+    private static Optional<EnsemblePlacementPolicyConfig> getEnsemblePlacementPolicyConfig(
+        Map<String, byte[]> customMetadata) {
+
+        byte[] ensemblePlacementPolicyConfigData = customMetadata.get(
+            EnsemblePlacementPolicyConfig.ENSEMBLE_PLACEMENT_POLICY_CONFIG);
+        if (ensemblePlacementPolicyConfigData != null) {
+            try {
+                return Optional.ofNullable(EnsemblePlacementPolicyConfig.decode(ensemblePlacementPolicyConfigData));
+            } catch (JsonUtil.ParseJsonException e) {
+                LOG.error("Failed to parse the ensemble placement policy config from the custom metadata", e);
+                return Optional.empty();
+            }
+        }
+        return Optional.empty();
+    }
+
+    private Map<String, List<String>> getDefaultIsolationGroups() {
+        Map<String, List<String>> isolationGroup = new HashMap<>();
+        isolationGroup.put(ISOLATION_BOOKIE_GROUPS, primaryIsolationGroups);
+        isolationGroup.put(SECONDARY_ISOLATION_BOOKIE_GROUPS, secondaryIsolationGroups);
+        return isolationGroup;
+    }
+
+    private static Map<String, List<String>> getIsolationGroup(EnsemblePlacementPolicyConfig ensemblePlacementPolicyConfig) {
```

Comment:
I think you can use `Pair<Set<String>, Set<String>>` instead of a HashMap. Because you only have two sets of groups, one is the primary group and the other one is a secondary group. A HashHamp introduces a lot of overheads.